packages/graphql-hooks/src/useClientRequest.js

```diff
@@ -96,63 +107,66 @@ function useClientRequest(query, initialOpts = {}) {
+        ...operation,
+        variables: revisedOpts.variables,
+        operationName: revisedOpts.operationName
+      }
 
-    const revisedCacheKey = client.getCacheKey(revisedOperation, revisedOpts)
+      const revisedCacheKey = client.getCacheKey(revisedOperation, revisedOpts)
 
-    // NOTE: There is a possibility of a race condition whereby
-    // the second query could finish before the first one, dispatching an old result
-    // see https://github.com/nearform/graphql-hooks/issues/150
-    activeCacheKey.current = revisedCacheKey
+      // NOTE: There is a possibility of a race condition whereby
+      // the second query could finish before the first one, dispatching an old result
+      // see https://github.com/nearform/graphql-hooks/issues/150
+      activeCacheKey.current = revisedCacheKey
 
-    const cacheHit =
-      revisedOpts.skipCache || !client.cache
-        ? null
-        : client.cache.get(revisedCacheKey)
+      const cacheHit =
+        revisedOpts.skipCache || !client.cache
+          ? null
+          : client.cache.get(revisedCacheKey)
 
-    if (cacheHit) {
-      dispatch({
-        type: actionTypes.CACHE_HIT,
-        result: cacheHit
-      })
+      if (cacheHit) {
+        dispatch({
+          type: actionTypes.CACHE_HIT,
+          result: cacheHit
+        })
 
-      return Promise.resolve(cacheHit)
-    }
+        return Promise.resolve(cacheHit)
+      }
 
-    dispatch({ type: actionTypes.LOADING })
-    return client.request(revisedOperation, revisedOpts).then(result => {
-      if (state.data && result.data && revisedOpts.updateData) {
-        if (typeof revisedOpts.updateData !== 'function') {
-          throw new Error('options.updateData must be a function')
+      dispatch({ type: actionTypes.LOADING })
+      return client.request(revisedOperation, revisedOpts).then(result => {
+        if (state.data && result.data && revisedOpts.updateData) {
+          if (typeof revisedOpts.updateData !== 'function') {
+            throw new Error('options.updateData must be a function')
+          }
+          result.data = revisedOpts.updateData(state.data, result.data)
         }
-        result.data = revisedOpts.updateData(state.data, result.data)
-      }
 
-      if (revisedOpts.useCache && client.cache) {
-        client.cache.set(revisedCacheKey, result)
-      }
+        if (revisedOpts.useCache && client.cache) {
+          client.cache.set(revisedCacheKey, result)
+        }
 
-      if (isMounted.current && revisedCacheKey === activeCacheKey.current) {
-        dispatch({
-          type: actionTypes.REQUEST_RESULT,
-          result
-        })
-      }
+        if (isMounted.current && revisedCacheKey === activeCacheKey.current) {
+          dispatch({
+            type: actionTypes.REQUEST_RESULT,
+            result
+          })
+        }
 
-      return result
-    })
-  }
+        return result
+      })
+    },
+    [client, initialOpts, operation, state.data]
```

Comment:
Good Idea!

Correct me if I'm wrong but is `state.data` changing after successful fetch?  If so deep equal check will fail and effect in https://github.com/nearform/graphql-hooks/issues/234 example will rerun.

Reply:
Yes, `state.data` will change, and it will cause `fetchData` to be re-created, but that is OK. `state.data` is a dependency of `fetchData` and it would be a mistake not to include it in `useCallback`'s dependencies. Regarding the example in #234, I think re-fetching could be avoided if you did things different. See my [comment](https://github.com/nearform/graphql-hooks/issues/234#issuecomment-500157999).

Reply:
So if it doesn't resolve the https://github.com/nearform/graphql-hooks/issues/234 , what problem it solves? Can you please please provide an example code where your memoization approach is beneficial? Thank you!

Reply:
Both of your arguments are valid that state is needed as a dependency and it does not fully solve the issue.

I have a suggestion here, let's remove the use of `state` from `fetchData`. We could pass `updateData` to reducer where state and `result.data` are available. That would eliminate the usage of state from `fetchData`.

For reducer:
```js
    case actionTypes.REQUEST_RESULT:
      return {
        ...action.result,
        data:
          action.updateData && state.data && action.result.data
            ? action.updateData(state.data, action.result.data)
            : state.data || action.result.data,
        cacheHit: false,
        loading: false
      }
```

in `fetchData` leaving the following bits:
```js
        if (
          revisedOpts.updateData &&
          typeof revisedOpts.updateData !== 'function'
        ) {
          throw new Error('options.updateData must be a function')
        }
```

```js
        if (isMounted.current && revisedCacheKey === activeCacheKey.current) {
          dispatch({
            type: actionTypes.REQUEST_RESULT,
            result,
            updateData: revisedOpts.updateData
          })
        }
```

@lynxtaa @olistic @bmullan91 what do you guys think?

Reply:
@edvinasbartkus Later we're storing updated result in cache:
```jsx
        result.data = revisedOpts.updateData(state.data, result.data)
      }

      if (revisedOpts.useCache && client.cache) {
        client.cache.set(revisedCacheKey, result)
      }
```

I'm not sure how it can be done with your approach.

The other way is to store the latest `state.data` in ref (like I do in https://github.com/nearform/graphql-hooks/pull/285).